### PR TITLE
Use `make_zero` and `Ref` inside `gradient`

### PR DESCRIPTION
**Pull request by mcabbott**
On Enzyme v0.11.20:
```julia
julia> using Enzyme, SparseArrays, StaticArrays

julia> gradient(Reverse, prod, [3.0 4.0])  # ok!
1×2 Matrix{Float64}:
 4.0  3.0

julia> gradient(Reverse, prod, @SArray [3.0 4.0])
1×2 SMatrix{1, 2, Float64, 2} with indices SOneTo(1)×SOneTo(2):
 0.0  0.0

julia> gradient(Reverse, sum, sparse([5.0 0.0 6.0]))
1×3 SparseMatrixCSC{Float64, Int64} with 0 stored entries:
  ⋅    ⋅    ⋅ 

julia> gradient(Reverse, abs2, 5.0)
0.0
```
With this PR:
```julia
julia> gradient(Reverse, prod, @SArray [3.0 4.0])
1×2 SMatrix{1, 2, Float64, 2} with indices SOneTo(1)×SOneTo(2):
 4.0  3.0

julia> gradient(Reverse, sum, sparse([5.0 0.0 6.0]))
1×3 SparseMatrixCSC{Float64, Int64} with 2 stored entries:
 1.0   ⋅   1.0

julia> gradient(Reverse, abs2, 5.0)
10.0
```
and also:
```julia
julia> xy = (x = [1.0, 2.0], y = [3.0, 4.0]);

julia> gradient(Reverse, z -> sum(z.x .* z.y), xy)
(x = [3.0, 4.0], y = [1.0, 2.0])

julia> xp1 = (x = [1.0, 2.0], p = 3);

julia> gradient(Reverse, z -> sum(z.x .^ z.p), xp1)
(x = [3.0, 12.0], p = 3)

julia> xp2 = (x = [1.0, 2.0], p = 3.0);

julia> gradient(Reverse, z -> sum(z.x .^ z.p), xp2)
ERROR: AssertionError: @NamedTuple{x::Vector{Float64}, p::Float64} has mixed internal activity types. See https://enzyme.mit.edu/julia/dev/#Mixed-Activity for more information
```
Edit: latest version changes the last example to:
```julia
julia> gradient(Reverse, z -> sum(z.x .^ z.p), xp2)
(x = [3.0, 12.0], p = 5.545177444479562)

julia> Zygote.gradient(z -> sum(z.x .^ z.p), xp2)[1]
(x = [3.0, 12.0], p = 5.545177444479562)
```

Not sure the test is in the right place. 

~~Should probably add the sparse and SVector cases as tests too, will try to figure out where.~~

### CODE REVIEW

File: src/Enzyme.jl [hunk 1]
```diff
@@ -916,9 +916,15 @@ grad = gradient(Reverse, f, [2.0, 3.0])
 ```
 """
 @inline function gradient(::ReverseMode, f, x)
-    dx = zero(x)
-    autodiff(Reverse, f, Active, Duplicated(x, dx))
-    dx
+    if Enzyme.Compiler.active_reg(typeof(x))
```
Comment (wsmoses) [inline src/Enzyme.jl:919]: Minor but I'm realizing the better check would be (saves compile time)

```julia
if active_reg_inner(Core.Typeof(f), #=seen=#(), #=world=#nothing, #=justActive=#Val(true)) == ActiveState
```

File: src/Enzyme.jl [hunk 2]
```diff
@@ -916,9 +916,15 @@ grad = gradient(Reverse, f, [2.0, 3.0])
...
+        dx = Ref(make_zero(x))
+        autodiff(Reverse, f∘only, Active, Duplicated(Ref(x), dx))
+        return only(dx)
+    else
+        dx = make_zero(x)
+        autodiff(Reverse, f, Active, Duplicated(x, dx))
+        return dx
+    end
```
Comment (wsmoses) [inline src/Enzyme.jl:927]: Also very minor, but while you're at it, do you mind adding the `x::Array` I mistakenly thought were on the other (in place, forward mode, etc) gradient wrapper codes?

Comment (mcabbott) [reply src/Enzyme.jl:927]: I suppose Array also forbids CuArray, although I'm not really sure what the status is there:
```jula
julia> using JLArrays  # fake GPU array

julia> x = jl([5.0 0.0 6.0])
1×3 JLArray{Float64, 2}:
 5.0  0.0  6.0

julia> gradient(Reverse, sum, x)
┌ Warning: TypeAnalysisDepthLimit
│ {[0]:Pointer, [0,0]:Pointer, [0,0,0]:Pointer, [0,0,0,-1]:Pointer, [0,0,0,0,0]:Pointer, [0,0,0,0,0,-1]:Integer, [0,0,0,0,8]:Integer, [0,0,0,0,9]:Integer, [0,0,0,0,10]:Integer, [0,0,0,0,11]:Integer, [0,0,0,0,12]:Integer, [0,0,0,0,13]:Integer, [0,0,0,0,14]:Integer, [0,0,0,0,15]:Integer, [0,0,0,0,16]:Integer, [0,0,0,0,17]:Integer, [0,0,0,0,18]:Integer, [0,0,0,0,19]:Integer, [0,0,0,0,20]:Integer, [0,0,0,0,21]:Integer, [0,0,0,0,22]:Integer, [0,0,0,0,23]:Integer, [0,0,0,0,24]:Integer, [0,0,0,0,25]:Integer, [0,0,0,0,26]:Integer, [0,0,0,0,27]:Integer, [0,0,0,0,28]:Integer, [0,0,0,0,29]:Integer, [0,0,0,0,30]:Integer, [0,0,0,0,31]:Integer, [0,0,0,0,32]:Integer, [0,0,0,0,33]:Integer, [0,0,0,0,34]:Integer, [0,0,0,0,35]:Integer, [0,0,0,0,36]:Integer, [0,0,0,0,37]:Integer, [0,0,0,0,38]:Integer, [0,0,0,0,39]:Integer, [0,0,0,16,-1]:Integer, [0,0,8]:Integer, [0,8]:Integer, [0,9]:Integer, [0,10]:Integer, [0,11]:Integer, [0,12]:Integer, [0,13]:Integer, [0,14]:Integer, [0,15]:Integer, [0,16]:Integer, [0,17]:Integer, [0,18]:Integer, [0,19]:Integer, [0,20]:Integer, [0,21]:Integer, [0,22]:Integer, [0,23]:Integer, [0,24]:Integer, [0,25]:Integer, [0,26]:Integer, [0,27]:Integer, [0,28]:Integer, [0,29]:Integer, [0,30]:Integer, [0,31]:Integer, [8]:Float@double}
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/U36Ed/src/utils.jl:59
┌ Warning: TypeAnalysisDepthLimit
│ {[0]:Pointer, [0,0]:Pointer, [0,0,0]:Pointer, [0,0,0,-1]:Pointer, [0,0,0,0,0]:Pointer, [0,0,0,0,0,-1]:Integer, [0,0,0,0,8]:Integer, [0,0,0,0,9]:Integer, [0,0,0,0,10]:Integer, [0,0,0,0,11]:Integer, [0,0,0,0,12]:Integer, [0,0,0,0,13]:Integer, [0,0,0,0,14]:Integer, [0,0,0,0,15]:Integer, [0,0,0,0,16]:Integer, [0,0,0,0,17]:Integer, [0,0,0,0,18]:Integer,
```

File: src/Enzyme.jl (hunk 1 above)
Comment (mcabbott) [reply src/Enzyme.jl]: I made this change, and it has the following effect, making my mixed activity namedtuple take the Ref path. I hope that sounds OK, the answer certainly looks right:
```julia
julia> xp2 = (x = [1.0, 2.0], p = 3.0);

julia> Enzyme.Compiler.active_reg(typeof(xp2))
ERROR: AssertionError: @NamedTuple{x::Vector{Float64}, p::Float64} has mixed internal activity types. See https://enzyme.mit.edu/julia/dev/#Mixed-Activity for more information

julia> Enzyme.Compiler.active_reg_inner(typeof(xp2), #=seen=#(), #=world=#nothing, #=justActive=#Val(true)) == Enzyme.Compiler.ActiveState
true

julia> gradient(Reverse, z -> sum(z.x .^ z.p), xp2)  # takes the Ref path
(x = [3.0, 12.0], p = 5.545177444479562)

julia> Zygote.gradient(z -> sum(z.x .^ z.p), xp2)[1]
(x = [3.0, 12.0], p = 5.545177444479562)
```
Maybe that needs a test.

Comment (wsmoses) [reply src/Enzyme.jl]: Yes it should be fine

File: src/Enzyme.jl (hunk 2 above)
Comment (wsmoses) [reply src/Enzyme.jl:940]: a docs PR I just made explains that a bit (https://github.com/EnzymeAD/Enzyme.jl/pull/1357).

I'm for the moment of the opinion that it's better to be conservative in throwing errors than risk accidentally running code which may get incorrect results/segfault.

```
[CUDA.jl](https://github.com/JuliaGPU/CUDA.jl) is only supported on Julia v1.7.0 and onwards. On v1.6, attempting to differentiate CUDA kernel functions will not use device overloads correctly and thus returns fundamentally wrong results.

Specifically, differentiating within device kernels is supported. See our [cuda tests](https://github.com/EnzymeAD/Enzyme.jl/blob/main/test/cuda.jl) for some examples. 

Differentiating through a heterogeneous (e.g. combined host and device) code presently requires defining a custom derivative that tells Enzyme that differentiating an `@cuda` call is done by performing `@cuda` of its generated derivative. For an example of this in Enzyme-C++ see [here](https://enzyme.mit.edu/getting_started/CUDAGuide/). Automating this for a better experience for CUDA.jl requires an update to [CUDA.jl](https://github.com/JuliaGPU/CUDA.jl/pull/1869/files), and is now available for Kernel Abstractions.

Differentiating host-side code when accesses device memory (e.g. `sum(CuArray)`) is not yet supported, but in progress.
```

File: docs/src/index.md [hunk 3]
```diff
@@ -202,3 +202,350 @@ julia> # Again, the optinal chunk size argument allows us to use vector forward
...
+If you pass in any temporary storage which may be involved in an active computation to a function you want to differentiate, you must also pass in a duplicated temporary storage for use in computing the derivatives. For example, consider the following function which uses a temporary buffer to compute the result.
+
+```jldoctest storage
+function f(x, tmp, k, n)
+    tmp[1] = 1.0
+    for i in 1:n
+        tmp[k] *= x
+    end
+    tmp[1]
+end
```
Comment (wsmoses) [inline docs/src/index.md:219]: LGTM except you'll need to rebase docs changes because they were moved to an FAQ page.

**Review by wsmoses — APPROVED** [review]

File: docs/src/index.md (hunk 3 above)
Comment (mcabbott) [reply docs/src/index.md]: Oh no, I messed that up... but now fixed.

**Comment by gdalle** [comment]
Can I allow myself a gentle bump for this PR, perhaps followed by a new release? It would be useful for the DifferentiationInterface.jl test suite with non-array objects

**Comment by mcabbott** [comment]
Re tagging a release, I have a draft of what I suggest here: https://github.com/EnzymeAD/Enzyme.jl/issues/1334#issuecomment-2016573247 

Since at present `gradient(Reverse, f, x)` is an error for complicated structs, perhaps any changes to what types it returns should be considered before tagging.

**Pull request merged**